### PR TITLE
HAAR-822 get child group details

### PR DESCRIPTION
**Pull request by AlTurner-MOJ**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt [hunk 1]
```diff
@@ -80,6 +91,24 @@ class HmppsManageUsersApiExceptionHandler {
...
+  fun handleValidationAnyException(e: HttpMessageNotReadableException): ResponseEntity<ErrorResponse?>? {
+    log.info("Validation exception: {}", e.message)
+    return ResponseEntity
+      .status(BAD_REQUEST)
+      .contentType(APPLICATION_JSON)
+      .body(ErrorResponse(status = BAD_REQUEST, userMessage = e.message, developerMessage = e.message))
+  }
+
+  @ExceptionHandler(MethodArgumentTypeMismatchException::class)
+  fun handleValidationAnyException(e: MethodArgumentTypeMismatchException): ResponseEntity<ErrorResponse?>? {
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt:104]: We usually match the function name to the exception thrown - although looks like existing line 113 also needs updating

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt [hunk 2]
```diff
@@ -73,6 +73,46 @@ class GroupsController(
...
+        content = [
+          Content(
+            mediaType = "application/json",
+            schema = Schema(implementation = ErrorResponse::class)
+          )
+        ]
+      )
+    ]
+  )
+  fun getChildGroupDetail(
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt:110]: General comment - not just yours - should we also have a 403 in the swagger? Maybe discuss as team.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt [hunk 3]
```diff
@@ -128,6 +129,13 @@ class ExternalUsersApiService(
       .bodyToMono(GroupDetails::class.java)
       .block(timeout) ?: throw GroupNotFoundException("get", group, "notfound")
 
+  fun getChildGroupDetail(group: String): ChildGroupDetails =
+    externalUsersWebClient.get()
+      .uri("/groups/child/$group")
+      .retrieve()
+      .bodyToMono(ChildGroupDetails::class.java)
+      .block(timeout) ?: throw RuntimeException("Failed to retrieve child group details")
+
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt:138]: Prob worth checking with Ravi what he's doing with the other timeouts for HAAR-931

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt [hunk 4]
```diff
@@ -1015,6 +1055,27 @@ class ExternalUsersApiMockServer : WireMockServer(WIREMOCK_PORT) {
...
   }
 
+  fun stubNotFound(url: String) {
+    stubFor(
+      get(urlEqualTo(url))
+        .willReturn(
+          aResponse()
+            .withHeader("Content-Type", "application/json")
+            .withStatus(HttpStatus.NOT_FOUND.value())
+            .withBody(
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt:1065]: Minor -HTTPStatus.NOT_FOUND

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt [hunk 5]
```diff
@@ -1015,6 +1055,27 @@ class ExternalUsersApiMockServer : WireMockServer(WIREMOCK_PORT) {
     )
   }
 
+  fun stubNotFound(url: String) {
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt:1058]: prob rename to stubGetNotFound - I think some of the PUTs return it too

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt [hunk 6]
```diff
@@ -155,6 +155,94 @@ class GroupsControllerIntTest : IntegrationTestBase() {
...
+        .expectStatus().isOk
+        .expectHeader().contentType(APPLICATION_JSON)
+        .expectBody()
+        .json(
+          """
+      {
+                    "groupCode": "CHILD_1",
+                    "groupName": "Child - Site 1 - Group 2"
+                  }  
+          """.trimIndent()
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt:197]: Does this formatting slightly?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt [hunk 7]
```diff
@@ -155,6 +155,94 @@ class GroupsControllerIntTest : IntegrationTestBase() {
...
+              "developerMessage" to "Access is denied",
+              "userMessage" to "Access is denied",
+            )
+          )
+        }
+    }
+
+    @Test
+    fun `returns error when child group not found`() {
+      externalUsersApiMockServer.stubNotFound("/groups/child/CHILD_1")
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt:224]: nice!

**Review by karenmillermoj — DISMISSED** [review]
lgtm

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt [hunk 8]
```diff
@@ -65,6 +67,15 @@ class HmppsManageUsersApiExceptionHandler {
...
 
+  @ExceptionHandler(RuntimeException::class)
+  fun handleException(e: RuntimeException): ResponseEntity<ErrorResponse> {
+    log.error("Unexpected exception", e)
+    return ResponseEntity
+      .status(INTERNAL_SERVER_ERROR)
+      .contentType(APPLICATION_JSON)
+      .body(ErrorResponse(status = INTERNAL_SERVER_ERROR.value(), developerMessage = e.message))
+  }
+
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt (hunk 3 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt:138]: Discussed, Ravi to test & feed back under his ticket

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt (hunk 4 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt (hunk 5 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt (hunk 6 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt:198]: Done

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**